### DISCUSSION TITLE
Is it possible to constrain spline shapes?

### DISCUSSION THREAD

**Discussion by iparperspective** [sdmTMB/sdmTMB#234] [General]
I have used squared terms and the scam package for R to force realistic covariate effect.

I wonder if sdmTMB is able to this? If not, would it be possible to implement it? I think that it would be an excellent addition.

Cheers,
Iosu

**Comment by ericward-noaa** [comment]
Yes -- by default, sdmTMB implements smooths as p-splines, rather than the tps approach that is the default in mgcv. This means there is a penalty on complexity -- with a goal of selecting more realistic shapes. You can also specify the knots (approx) as 

`density ~ s(year, k = 10) `

With smaller values of `k` leading to more smooth relationships. If you wanted a low order polynomial, k=3 or 4 is probably reasonable.

**Reply by iparperspective** [reply]
I see, thanks Eric. 
And would it be possible to force concave shapes (unimodal shapes)

**Reply by ericward-noaa** [reply]
Currently, it's not possible with using smooths in sdmTMB. I don't think it's an option in `brms` either -- 

https://discourse.mc-stan.org/t/monotonic-shape-constrained-splines-still-possible-in-brms/13117/5

But one could set up a model in sdmTMB with quadratic coeffcients and put a very strong prior on the coefficients -- this wouldn't exactly force the shape, but may help to constrain things
